File: src/api/events.ts
/**
 * Tauri event 订阅工具。
 *
 * 事件名与 src-tauri/src/events.rs 中的常量对应。
 */

import { listen, type UnlistenFn } from "@tauri-apps/api/event";

import type {
  Alert,
  IntradayPoint,
  KlinePoint,
  Quote,
  Subscription,
} from "./tauri";

export const Events = {
  QuoteUpdated: "quote:updated",
  IntradayUpdated: "intraday:updated",
  KlineUpdated: "kline:updated",
  SubscriptionChanged: "subscription:changed",
  PollError: "poll:error",
  AlertNew: "alert:new",
} as const;

export interface QuoteUpdatedPayload {
  symbol: string;
  quote: Quote;
}
export interface IntradayUpdatedPayload {
  symbol: string;
  points: IntradayPoint[];
}
export interface KlineUpdatedPayload {
  symbol: string;
  period: string;
  points: KlinePoint[];
}
export interface SubscriptionChangedPayload {
  subscriptions: Subscription[];
}
export interface PollErrorPayload {
  symbol: string;
  kind: string;
  error: string;
}

export interface AlertNewPayload {
  alerts: Alert[];
}

export function onQuoteUpdated(
  cb: (p: QuoteUpdatedPayload) => void,
): Promise<UnlistenFn> {
  return listen<QuoteUpdatedPayload>(Events.QuoteUpdated, (e) => cb(e.payload));
}

export function onIntradayUpdated(
  cb: (p: IntradayUpdatedPayload) => void,
): Promise<UnlistenFn> {
  return listen<IntradayUpdatedPayload>(Events.IntradayUpdated, (e) =>
    cb(e.payload),
  );
}

export function onKlineUpdated(
  cb: (p: KlineUpdatedPayload) => void,
): Promise<UnlistenFn> {
  return listen<KlineUpdatedPayload>(Events.KlineUpdated, (e) => cb(e.payload));
}

export function onSubscriptionChanged(
  cb: (p: SubscriptionChangedPayload) => void,
): Promise<UnlistenFn> {
  return listen<SubscriptionChangedPayload>(Events.SubscriptionChanged, (e) =>
    cb(e.payload),
  );
}

export function onPollError(
  cb: (p: PollErrorPayload) => void,
): Promise<UnlistenFn> {
  return listen<PollErrorPayload>(Events.PollError, (e) => cb(e.payload));
}

export function onAlertNew(
  cb: (p: AlertNewPayload) => void,
): Promise<UnlistenFn> {
  return listen<AlertNewPayload>(Events.AlertNew, (e) => cb(e.payload));
}

File: src-tauri/src/events.rs
//! Tauri 事件常量与 payload 类型。前端通过 `listen` 订阅这些事件名。

use serde::Serialize;

use crate::models::{Alert, IntradayPoint, KlinePoint, Quote, Subscription};

pub const EV_QUOTE_UPDATED: &str = "quote:updated";
pub const EV_INTRADAY_UPDATED: &str = "intraday:updated";
pub const EV_KLINE_UPDATED: &str = "kline:updated";
pub const EV_SUBSCRIPTION_CHANGED: &str = "subscription:changed";
pub const EV_POLL_ERROR: &str = "poll:error";
pub const EV_ALERT_NEW: &str = "alert:new";

#[derive(Debug, Clone, Serialize)]
pub struct QuoteUpdated {
    pub symbol: String,
    pub quote: Quote,
}

#[derive(Debug, Clone, Serialize)]
pub struct IntradayUpdated {
    pub symbol: String,
    pub points: Vec<IntradayPoint>,
}

#[derive(Debug, Clone, Serialize)]
pub struct KlineUpdated {
    pub symbol: String,
    pub period: String,
    pub points: Vec<KlinePoint>,
}

#[derive(Debug, Clone, Serialize)]
pub struct SubscriptionChanged {
    pub subscriptions: Vec<Subscription>,
}

#[derive(Debug, Clone, Serialize)]
pub struct PollError {
    pub symbol: String,
    pub kind: String,
    pub error: String,
}

/// 新增告警事件（单条或批量）
#[derive(Debug, Clone, Serialize)]
pub struct AlertNew {
    pub alerts: Vec<Alert>,
}
